### PR TITLE
 AP_BoardConfig: add targetted check for cube black internal sensors

### PR DESCRIPTION
**Pull request by jschall**
Since MS5611 irritatingly doesn't have a WHOAMI register, I had to do a fairly complicated CRC check on its PROM.

### CODE REVIEW

File: libraries/AP_BoardConfig/board_drivers.cpp [hunk 1]
```diff
@@ -152,6 +152,90 @@ bool AP_BoardConfig::spi_check_register(const char *devname, uint8_t regnum, uin
...
+        }
+    }
+
+    return (n_rem >> 12) & 0xF;
+}
+
+static bool check_ms5611(const char* devname) {
+    auto dev = hal.spi->get_device(devname);
+    AP_HAL::Semaphore *dev_sem = dev->get_semaphore();
+    if (!dev || !dev_sem) {
```
Comment (WickedShell) [inline libraries/AP_BoardConfig/board_drivers.cpp]: You've already dereferenced `dev`, defeating the nullptr check here.

File: libraries/AP_BoardConfig/board_drivers.cpp [hunk 2]
```diff
@@ -199,11 +285,34 @@ void AP_BoardConfig::validate_board_type(void)
...
 
+
+
 /*
   auto-detect board type
  */
 void AP_BoardConfig::board_autodetect(void)
 {
+#if defined(HAL_CHIBIOS_ARCH_CUBEBLACK)
+    {
```
Comment (WickedShell) [inline libraries/AP_BoardConfig/board_drivers.cpp]: Nitpick: Why the extra scoping here?

File: libraries/AP_BoardConfig/board_drivers.cpp [hunk 3]
```diff
@@ -199,11 +285,34 @@ void AP_BoardConfig::validate_board_type(void)
...
+        if (state.board_type != PX4_BOARD_PIXHAWK2) {
+            state.board_type.set(PX4_BOARD_PIXHAWK2);
+        }
+
+        bool success = true;
+        if (!spi_check_register("mpu9250", MPUREG_WHOAMI, MPU_WHOAMI_MPU9250)) { success = false; }
```
Comment (WickedShell) [inline libraries/AP_BoardConfig/board_drivers.cpp]: I'm okay with it in this form, but ideally I think we'd provide this via a set of entries in the HWDEF.

File: libraries/AP_BoardConfig/board_drivers.cpp (hunk 2 above)
Comment (jschall) [reply libraries/AP_BoardConfig/board_drivers.cpp]: Scope is for the success variable.

File: libraries/AP_BoardConfig/board_drivers.cpp [hunk 4]
```diff
@@ -152,6 +152,90 @@ bool AP_BoardConfig::spi_check_register(const char *devname, uint8_t regnum, uin
     return v == value;
 }
 
+/**
+ * MS56XX crc4 method from datasheet for 16 bytes (8 short values)
+ */
+static uint16_t crc4(uint16_t *data)
```
Comment (tridge) [inline libraries/AP_BoardConfig/board_drivers.cpp]: I'd rather make the functions in AP_Baro_MS5611 be public static and call them, or move the crc function to AP_Math/crc.c with suitable name

File: libraries/AP_BoardConfig/board_drivers.cpp [hunk 5]
```diff
@@ -152,6 +152,90 @@ bool AP_BoardConfig::spi_check_register(const char *devname, uint8_t regnum, uin
...
+
+    static const uint8_t CMD_MS56XX_RESET = 0x1E;
+    static const uint8_t CMD_MS56XX_PROM = 0xA0;
+
+    dev->transfer(&CMD_MS56XX_RESET, 1, nullptr, 0);
+    hal.scheduler->delay(4);
+
+    uint16_t prom[8];
+    bool all_zero = true;
+    for (uint8_t i = 0; i < 8; i++) {
```
Comment (tridge) [inline libraries/AP_BoardConfig/board_drivers.cpp]: I think we can do most of this via exposing static functions in AP_Baro_MS5611

File: libraries/AP_BoardConfig/board_drivers.cpp (hunk 2 above)
Comment (tridge) [reply libraries/AP_BoardConfig/board_drivers.cpp]: I think adding a board_check_CubeBlack which gets called from here would be better. This won't be the last board where we hit this type of issue

Comment (jschall) [reply libraries/AP_BoardConfig/board_drivers.cpp]: Took tridge's suggestion

File: libraries/AP_BoardConfig/board_drivers.cpp (hunk 1 above)
Comment (jschall) [reply libraries/AP_BoardConfig/board_drivers.cpp]: Fixed.

File: libraries/AP_BoardConfig/board_drivers.cpp (hunk 3 above)
Comment (jschall) [reply libraries/AP_BoardConfig/board_drivers.cpp]: Agreed.

File: libraries/AP_BoardConfig/board_drivers.cpp (hunk 5 above)
Comment (jschall) [reply libraries/AP_BoardConfig/board_drivers.cpp]: I did the CRC4 function, but I'd prefer to just leave this as-is.

**Comment by proficnc** [comment]
Thanks all

**Pull request merged**